### PR TITLE
cephfs-top: make cephfs-top display scrollable

### PR DESCRIPTION
**Pull request by joscollin**
Fixes: https://tracker.ceph.com/issues/55197

* Implements Vertical and Horizontal scrolling of cephfs-top display. Use `Up/Down` Arrow Keys, `PgUp/PgDn` Keys or `mouse` for Vertical scrolling. Use `Left/Right` Arrow Keys and `Home/End` keys for Horizontal scrolling.
* Maintains the `run_display` and `run_all_display` applications separately for customising, if needed. 
* Avoids the horizontal length (hlen) calculations of screen width. From now onwards, add n number of columns/fields and just Increase the `PAD_WIDTH` appropriately. Don't calculate the horizontal length anymore.
* This code currently supports a `PAD_HEIGHT` of 10,000, which means I'm expecting nearly 10,000 clients only. We could make it configurable or calculate it either (later). The important thing to notice is that the `PAD_HEIGHT` and `PAD_WIDTH` are constant with in the run display application. It could be configured, but it's not meant for changing dynamically. It's not meant for resizing and recreating dynamically. Create the pad once and use it until the end of the application. Only a **portion of the viewport** (clipping box) is refreshed at a time.
* The cephfs-top new UI with scrolling displays best in a maximised terminal window. It works in a tiny sized terminal window too, but the user has to give enough window size for him to see it fully functioning.

To test this well, try great number of clients like:
```
$ for i in {1..175}; do sudo mkdir -p /mnt/cephfs$i/; done
$ for i in {1..100}; do sudo ./bin/ceph-fuse -m <ip>:<port>  /mnt/cephfs$i/; done
$ for i in {101..150}; do sudo ./bin/ceph-fuse --client_mds_namespace fs1 -m <ip>:<port>  /mnt/cephfs$i/; done
$ for i in {151..175}; do sudo ./bin/ceph-fuse --client_mds_namespace fs2 -m <ip>:<port>  /mnt/cephfs$i/; done
```

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/tools/cephfs/top/cephfs-top
Comment (neesingh-rh) [inline src/tools/cephfs/top/cephfs-top:660]: Currently, we are having a hardcoded value of `self.PAD_HEIGHT` due to which on vertical scrolling, we keep going until we reach `self.PAD_HEIGHT` value, even when we have very less metrics to display. We can add something like `self.PAD_HEIGHT = self.tablehead_y + self.header_height` before `line:660` to achieve a self adjusting `self.PAD_HEIGHT` value which will allow us to scroll till our metrics display ends.Similarly, something like this for `run_display`.I can see, its mentioned in desc. that we configure it later, but I don't think we have something to wait for this.It will make the scroll feature better.

File: src/tools/cephfs/top/cephfs-top
Comment (neesingh-rh) [inline src/tools/cephfs/top/cephfs-top:667]: I can scroll `UP/DOWN` using `KEYS` as well as using two fingers on touchpad, but couldn't scroll `left/right` using two fingers.I am not sure if this can be implemented or not, but we used to have both functioning wherever scroll thing works.It will good to have it also.Just an opinion.

File: src/tools/cephfs/top/cephfs-top
Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:661]: I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.

That said, this issue is already fixed in the latest code. You should checkout the latest code.

File: src/tools/cephfs/top/cephfs-top
Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:680]: I haven't considered enabling the horizontal mouse scroll yet, as there is not much need for it. cephfs-top has fixed number (20) of columns in the table, which doesn't seem difficult to scroll using the arrow keys.  An appropriate terminal font size would fit the entire table too. The vertical scrolling has much more importance because of the greater/unknown number of items to accommodate and scroll.

File: src/tools/cephfs/top/cephfs-top
Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:661]: I checked the latest code, the above mentioned issue is fixed.But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below a specific window size.

Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:661]: > I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.
> 
> That said, this issue is already fixed in the latest code. You should checkout the latest code.

We can handle the pad height separately inside both display functions for the comparison in `KEY_UP` and `KEY_DOWN` instead of using the `self.PAD_HEIGHT`, with reference to the way it was done before latest change.

File: src/tools/cephfs/top/cephfs-top
Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:680]: Yeah, it's not much needed.We can ignore.

File: src/tools/cephfs/top/cephfs-top
Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:661]: > > I don't think adjusting the pad height every time is a good idea to have in a user interface. It could be configured, but not adjustable based on the items to be displayed which is what we were doing in the old cephfs-top.
> > That said, this issue is already fixed in the latest code. You should checkout the latest code.
> 
> We can handle the pad height separately inside both display functions for the comparison in `KEY_UP` and `KEY_DOWN` instead of using the `self.PAD_HEIGHT`, with reference to the way it was done before latest change.

@neesingh-rh 
I request you to understand what's the need for a pad and how it differs from a window. You should try to understand why a pad height is not meant to adjust dynamically and create a new pad every time we get new metrics. Scrolling is not about adjusting the pad height or pad width. Then what is it about? Please understand that part.

> But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below , at a specific window size.

Well, the user should have a sensible screen size first, before asking why it's not showing the last few rows. I meant, what's the use case of having such a tiny size window? And what prevents the user to drag and increase the window size a little bit? In real-time, the user would have larger window sizes instead. I've checked out this particular issue and this happens only when the window size is tiny, in which curses doesn't have enough room to make a scrolling-end action. In fact, it does it internally. But the user doesn't allow himself to see it.

Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:661]: > > But, there are cases when we aren't able to see all the filesystem metrics as scroll doesn't allow to move below , at a specific window size.
> 
> Well, the user should have a sensible screen size first, before asking why it's not showing the last few rows. I meant, what's the use case of having such a tiny size window? And what prevents the user to drag and increase the window size a little bit? In real-time, the user would have larger window sizes instead. I've checked out this particular issue and this happens only when the window size is tiny, in which curses doesn't have enough room to make a scrolling-end action. In fact, it does it internally. But the user doesn't allow himself to see it.

Yeah, Make sense. I checked again, this issue comes into picture only when there is a tiny window size.Otherwise, its working fine.

Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:661]: It works with tiny sized window too. But if the user doesn't give enough viewport size to see the scroll-end action, he can't see it ends. There's no shrink and show mechanism involved and we can't use a different offset increment value as the viewport shrinks.

File: src/tools/cephfs/top/cephfs-top
Comment (neesingh-rh) [inline src/tools/cephfs/top/cephfs-top:596]: @vshankar IIRC, you suggested to have the fields name along with the filesystem name line should be fixed while scrolling so that user can have the glance of which filesystem metrics he's looking currently and can look the fields easily.If so, then currently we are not having the filesystem names lines fixed on scrolling.See,
Otherwise, rest looks good to me.

Comment (vshankar) [reply src/tools/cephfs/top/cephfs-top:596]: Wasn't that related to @joscollin changes that introduce scrolling/pagination?

Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:596]: I am talking about the suggestion in the watercooler for scrolling, if not wrong.

Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:596]: There are no qt kind of widgets in curses to achieve such things. The text displayed in the fsstats pad is pre-loadad as a whole. The current text should be erased in-order to display the following text at the same position while scrolling. In other words, each scroll key press refreshes different portion of the pad (which is already text loaded), which creates the illusion of scrolling. So there's no way to stop the movement of `Filesystem;` name line, which is different for each fs object in the code. In other words, curses doesn't identify it as a special line (or object) to impose some restrictions to it. The top line moving was stopped, as it's common text for all filesystems. So I've drawn it only at the top and refreshing from the next line onwards. Hope that clarifies things.

Comment (vshankar) [reply src/tools/cephfs/top/cephfs-top:596]: > @vshankar IIRC, you suggested to have the fields name along with the filesystem name line should be fixed while scrolling so that user can have the glance of which filesystem metrics he's looking currently and can look the fields easily.If so, then currently we are not having the filesystem names lines fixed on scrolling.See, [5493030](https://github.com/ceph/ceph/commit/549303071d7d8b77abeae856cfb89a2f7378a711) Otherwise, rest looks good to me.

OK. What do you mean by "not having filesystem names lines fixed on scrolling"?

Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:596]: He meant something shown in this video: https://www.youtube.com/watch?v=6zS0OavqbAc. Watch at 0:40.

Comment (neesingh-rh) [reply src/tools/cephfs/top/cephfs-top:596]: > > @vshankar IIRC, you suggested to have the fields name along with the filesystem name line should be fixed while scrolling so that user can have the glance of which filesystem metrics he's looking currently and can look the fields easily.If so, then currently we are not having the filesystem names lines fixed on scrolling.See, [5493030](https://github.com/ceph/ceph/commit/549303071d7d8b77abeae856cfb89a2f7378a711) Otherwise, rest looks good to me.
> 
> OK. What do you mean by "not having filesystem names lines fixed on scrolling"?

I meant to say if we scroll now, the `TOP_LINE` (the line where the names of fields are displayed ) is always visible but the line `Filesystem: cephfs - 1 client(s)` is being scrolled along with the metrics too.
As per @joscollin reply and my findings, I can say its not easy to have that in ncurses. But I am unsure whether we can't implement. But that's only w.r.t the suggestion, it looks fine even without that.

**Comment by joscollin** [comment]
jenkins test api

**Comment by joscollin** [comment]
Test passed:
http://pulpito.front.sepia.ceph.com/jcollin-2022-10-24_10:06:27-fs-wip-F55197-cephfs-top-scroll-distro-default-smithi/

File: src/tools/cephfs/top/cephfs-top
Comment (joscollin) [reply src/tools/cephfs/top/cephfs-top:596]: @neesingh-rh The content changes are not within the scope of this PR. Only scrolling fixes are in scope. But I've still incorporated the sensible things mentioned in the WC, especially making the table header static and it works very well.

**Comment by joscollin** [comment]
jenkins test api

**Comment by tchaikov** [comment]
jenkins test api

**Comment by joscollin** [comment]
jenkins test windows

**Comment by joscollin** [comment]
jenkins test windows

**Comment by vshankar** [comment]
Waiting for an ACK from @neesingh-rh for merge.

**Review by neesingh-rh — APPROVED** [review]

**Comment by vshankar** [comment]
> > Waiting for an ACK from @neesingh-rh for merge.
> 
> Now this delay blocks me from creating a new PR (rebased). @vshankar Can you approve instead?

Well ok :)

**Pull request merged**
